ext/dba/dba.c

```diff
@@ -525,42 +561,45 @@ static void php_dba_open(INTERNAL_FUNCTION_PARAMETERS, bool persistent)
 		RETURN_THROWS();
 	}
 
+	char *resource_key;
+	size_t resource_key_len = spprintf(&resource_key, 0,
+		"dba_%d_%s_%s_%s", persistent, ZSTR_VAL(path), ZSTR_VAL(mode), handler_str ? ZSTR_VAL(handler_str) : ""
+	);
+
 	if (persistent) {
 		zend_resource *le;
 
-		if (handler_str) {
-			persistent_resource_key = zend_string_concat3(
-				ZSTR_VAL(path), ZSTR_LEN(path),
-				ZSTR_VAL(mode), ZSTR_LEN(mode),
-				ZSTR_VAL(handler_str), ZSTR_LEN(handler_str)
-			);
-		} else {
-			persistent_resource_key = zend_string_concat2(
-				ZSTR_VAL(path), ZSTR_LEN(path),
-				ZSTR_VAL(mode), ZSTR_LEN(mode)
-			);
-		}
-
 		/* try to find if we already have this link in our persistent list */
-		if ((le = zend_hash_find_ptr(&EG(persistent_list), persistent_resource_key)) != NULL) {
-			FREE_PERSISTENT_RESOURCE_KEY();
+		if ((le = zend_hash_str_find_ptr(&EG(persistent_list), resource_key, resource_key_len)) != NULL) {
 			if (le->type != le_pdb) {
 				// TODO This should never happen
+				FREE_RESOURCE_KEY();
 				RETURN_FALSE;
 			}
 
-			info = (dba_info *)le->ptr;
+			object_init_ex(return_value, dba_connection_ce);
+			dba_connection *connection = Z_DBA_CONNECTION_P(return_value);
+			connection->info = (dba_info *)le->ptr;
+			connection->hash = zend_string_init(resource_key, resource_key_len, persistent);
+			if (persistent) {
+				GC_MAKE_PERSISTENT_LOCAL(connection->hash);
+			}
 
-			GC_ADDREF(le);
-			RETURN_RES(zend_register_resource(info, le_pdb));
+			if (zend_hash_exists(&DBA_G(connections), connection->hash)) {
+				zend_hash_del(&DBA_G(connections), connection->hash);
+			}
+
+			zend_hash_add_new(&DBA_G(connections), connection->hash, return_value);
+			FREE_RESOURCE_KEY();
+			return;
```

Comment:
I'm confused here, can't we get rid of the persistent resource by storing the pointer for the DBH in the zval instead of the constructed object?